rules/coding-style/src/Rector/Throw_/AnnotateThrowablesRector.php

```diff
@@ -173,83 +211,210 @@ private function identifyThrownThrowablesInStaticCall(StaticCall $staticCall): a
+
+            if ($throwsType instanceof FullyQualifiedObjectType) {
+                $thrownClass = $throwsType->getClassName();
+            }
+
+            if ($thrownClass !== null) {
+                $alreadyAnnotatedThrowables[] = $thrownClass;
+            }
+        }
+
+        return $alreadyAnnotatedThrowables;
+    }
+
+    private function compareThrowables(array $foundThrownThrowables, array $alreadyAnnotatedThrowables): void
+    {
+        $normalizeNamespace = static function (string $class): string {
+            $class = ltrim($class, '\\');
+            return '\\' . $class;
+        };
+
+        $foundThrownThrowables = array_map($normalizeNamespace, $foundThrownThrowables);
+        $alreadyAnnotatedThrowables = array_map($normalizeNamespace, $alreadyAnnotatedThrowables);
+
+        $filterClasses = static function (string $class): bool {
+            return class_exists($class) || interface_exists($class);
+        };
+
+        $foundThrownThrowables = array_filter($foundThrownThrowables, $filterClasses);
+        $alreadyAnnotatedThrowables = array_filter($alreadyAnnotatedThrowables, $filterClasses);
+
+        foreach ($foundThrownThrowables as $foundThrownThrowable) {
+            if (! in_array($foundThrownThrowable, $alreadyAnnotatedThrowables, true)) {
+                $this->throwablesToAnnotate[] = $foundThrownThrowable;
+            }
+        }
     }
 
     private function buildThrowsDocComment(string $throwableClass): AttributeAwarePhpDocTagNode
     {
-        $genericTagValueNode = new ThrowsTagValueNode(new IdentifierTypeNode('\\' . $throwableClass), '');
+        $genericTagValueNode = new ThrowsTagValueNode(new IdentifierTypeNode($throwableClass), '');
 
         return new AttributeAwarePhpDocTagNode('@throws', $genericTagValueNode);
     }
 
-    private function buildFQN(Throw_ $throw): ?string
+    private function identifyCallee(Node $node): ?Node
     {
-        if (! $throw->expr instanceof New_) {
+        $callee = $node->getAttribute('previousExpression');
+        while (true) {
+            if (
+                $callee instanceof ClassMethod ||
+                $callee instanceof Function_
+            ) {
+                break;
+            }
+
+            $callee = $callee->getAttribute('previousExpression');
+        }
+
+        return $callee;
+    }
+
+    private function expandClassNamesAnnotatedInFunction(
+        ReflectionFunction $reflectionFunction,
+        array $classNames
+    ): array {
+        $functionNode = $this->reflectionFunctionGetNode($reflectionFunction);
+        if (! $functionNode instanceof Namespace_) {
+            return [];
+        }
+
+        $uses = $this->getUses($functionNode);
+
+        $expandedClasses = [];
+        foreach ($classNames as $className) {
+            $shortClassName = $this->getShortName($className);
+            $expandedClasses[] = $uses[$shortClassName] ?? $className;
+        }
+
+        return $expandedClasses;
+    }
```

Comment:
This has potential for standalone service, e.g. FunctionAnnotationExpander

Reply:
Namespace?

Reply:
`AnnotationReflection`

Reply:
In which folder do I put the file?

Reply:
With PSR-4, namespace == directory

so `src/Directory/...`